File: apps/console/lib/announcements.test.ts
import { describe, expect, it } from "vitest";
import { isLive, state, targeting, type Announcement } from "./announcements";

const base: Announcement = {
  id: "a1", title: "t", body: "b", severity: "info",
  starts_at: "2026-09-01T00:00:00Z", ends_at: null,
  audience_filter: {}, is_published: true, updated_at: "2026-09-01T00:00:00Z",
};
const NOW = new Date("2026-09-05T12:00:00Z");

describe("state", () => {
  it("separates the four states an operator scans for", () => {
    expect(state({ ...base, is_published: false }, NOW)).toBe("draft");
    expect(state({ ...base, starts_at: "2026-09-09T00:00:00Z" }, NOW)).toBe("scheduled");
    expect(state({ ...base, ends_at: "2026-09-02T00:00:00Z" }, NOW)).toBe("ended");
    expect(state(base, NOW)).toBe("live");
  });

  it("calls an unpublished future announcement a draft, not scheduled", () => {
    // Scheduling something that was never published would read as "this will
    // go out on Tuesday" when nothing will.
    expect(state({ ...base, is_published: false, starts_at: "2026-09-09T00:00:00Z" }, NOW)).toBe("draft");
  });
});

describe("isLive", () => {
  it("agrees with the API's three conditions", () => {
    expect(isLive(base, NOW)).toBe(true);
    expect(isLive({ ...base, is_published: false }, NOW)).toBe(false);
    expect(isLive({ ...base, starts_at: "2026-09-09T00:00:00Z" }, NOW)).toBe(false);
    expect(isLive({ ...base, ends_at: "2026-09-02T00:00:00Z" }, NOW)).toBe(false);
  });

  it("treats a null end as running indefinitely rather than as ended", () => {
    expect(isLive({ ...base, ends_at: null }, NOW)).toBe(true);
  });
});

describe("targeting", () => {
  it("reads the two keys the query matches on", () => {
    expect(targeting({ products: ["mark8ly"], statuses: ["active"] }))
      .toEqual({ products: ["mark8ly"], statuses: ["active"] });
  });

  it("reads an absent key as untargeted, matching the query's NULL branch", () => {
    // `audience_filter->'products' IS NULL` means EVERY product. Reading it as
    // "no products" would render an estate-wide broadcast as reaching nobody.
    expect(targeting({})).toEqual({ products: [], statuses: [] });
  });

  it("ignores a filter key it does not know rather than failing", () => {
    // The schema is "intentionally permissive so we can grow filters without a
    // migration"; a console that threw on an unknown key would make growing
    // one a breaking change.
    expect(targeting({ products: ["mark8ly"], tags: ["beta"] }).products).toEqual(["mark8ly"]);
  });

  it("drops non-string entries rather than rendering them", () => {
    expect(targeting({ products: ["mark8ly", 42, null] }).products).toEqual(["mark8ly"]);
  });
});

File: platform-api/internal/modules/announcements/internal/repository/announcements_test.go
package repository_test

import (
	"context"
	"testing"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/tesserix/tesserix-home/platform-api/internal/modules/announcements/internal/repository"
	"github.com/tesserix/tesserix-home/platform-api/internal/platform/testdb"
)

// The targeting is a JSONB containment query with two NULL branches, and the
// branches are the part worth pinning: `audience_filter->'products'` ABSENT
// means "every product", not "no product". Dropping either branch hides every
// untargeted broadcast — which is most of them — and the symptom is an empty
// banner rather than an error.

func seed(t *testing.T, pool *pgxpool.Pool, title, filter string, published bool, window string) {
	t.Helper()
	starts, ends := "now() - interval '1 hour'", "NULL"
	switch window {
	case "future":
		starts = "now() + interval '1 day'"
	case "ended":
		ends = "now() - interval '1 minute'"
	}
	_, err := pool.Exec(context.Background(),
		`INSERT INTO platform_announcements (title, body, severity, audience_filter, is_published, starts_at, ends_at)
		 VALUES ($1, 'body', 'info', $2::jsonb, $3, `+starts+`, `+ends+`)`,
		title, filter, published)
	if err != nil {
		t.Fatalf("seeding %q: %v", title, err)
	}
}

func activeTitles(t *testing.T, pool *pgxpool.Pool, product, status string) map[string]bool {
	t.Helper()
	got, err := repository.Active(context.Background(), pool, product, status)
	if err != nil {
		t.Fatalf("Active: %v", err)
	}
	seen := map[string]bool{}
	for _, a := range got {
		seen[a.Title] = true
	}
	return seen
}

func TestAnUntargetedAnnouncementReachesEveryProduct(t *testing.T) {
	pool := testdb.New(t)
	seed(t, pool, "everyone", `{}`, true, "live")

	if !activeTitles(t, pool, "mark8ly", "active")["everyone"] {
		t.Error("an announcement with no product filter did not reach mark8ly — the NULL branch is what makes an untargeted broadcast universal")
	}
	if !activeTitles(t, pool, "kora", "active")["everyone"] {
		t.Error("the same announcement did not reach kora")
	}
}

func TestAProductTargetedAnnouncementReachesOnlyThatProduct(t *testing.T) {
	pool := testdb.New(t)
	seed(t, pool, "mark8ly-only", `{"products":["mark8ly"]}`, true, "live")

	if !activeTitles(t, pool, "mark8ly", "active")["mark8ly-only"] {
		t.Error("a mark8ly-targeted announcement did not reach mark8ly")
	}
	if activeTitles(t, pool, "kora", "active")["mark8ly-only"] {
		t.Error("a mark8ly-targeted announcement reached kora")
	}
}

func TestStatusTargetingFiltersTheSameWay(t *testing.T) {
	pool := testdb.New(t)
	seed(t, pool, "trialing-only", `{"statuses":["trialing"]}`, true, "live")

	if !activeTitles(t, pool, "mark8ly", "trialing")["trialing-only"] {
		t.Error("a trialing-targeted announcement did not reach a trialing tenant")
	}
	if activeTitles(t, pool, "mark8ly", "active")["trialing-only"] {
		t.Error("a trialing-targeted announcement reached an active tenant")
	}
}

func TestUnpublishedAndOutOfWindowAreNotServed(t *testing.T) {
	// Three ways an announcement is not live. A merchant seeing a draft, or a
	// maintenance notice for work that finished, is worse than seeing nothing.
	pool := testdb.New(t)
	seed(t, pool, "draft", `{}`, false, "live")
	seed(t, pool, "scheduled", `{}`, true, "future")
	seed(t, pool, "expired", `{}`, true, "ended")
	seed(t, pool, "live-one", `{}`, true, "live")

	got := activeTitles(t, pool, "mark8ly", "active")
	for _, hidden := range []string{"draft", "scheduled", "expired"} {
		if got[hidden] {
			t.Errorf("%q was served", hidden)
		}
	}
	if !got["live-one"] {
		t.Error("the live announcement was not served — the filters exclude too much")
	}
}

func TestAnEmptyResultIsAListAndNotNull(t *testing.T) {
	pool := testdb.New(t)
	seed(t, pool, "kora-only", `{"products":["kora"]}`, true, "live")

	got, err := repository.Active(context.Background(), pool, "mark8ly", "active")
	if err != nil {
		t.Fatalf("Active: %v", err)
	}
	if got == nil {
		t.Error("an empty result was nil — it marshals as null, and a client expecting a list would be reading a bug in this API")
	}
	if len(got) != 0 {
		t.Errorf("got %d announcements, want none", len(got))
	}
}
